File: AdSwitcher-iOS/AdSwitcher/Adapters/VungleAdapter/VungleAdapter.m
//
//  VungleAdapter.m
//  AdSwitcher
//
//  Created by tkyaji on 2016/07/01.
//  Copyright © 2017年 adwitcher. All rights reserved.
//

#import "VungleAdapter.h"

@implementation VungleAdapter {
    UIViewController *_viewController;
}

@synthesize interstitialAdDelegate;

#pragma - InterstitialAdAdapter

- (void)interstitialAdInitialize:(UIViewController *)viewController parameters:(NSDictionary<NSString *,NSString *> *)parameters testMode:(BOOL)testMode {
    _viewController = viewController;
    NSString *appId = [parameters objectForKey:@"app_id"];
    
    [[VungleSDK sharedSDK] setDelegate:self];
    [[VungleSDK sharedSDK] setLoggingEnabled:testMode];
    [[VungleSDK sharedSDK] startWithAppId:appId];
}

- (void)interstitialAdLoad {
    _DLOG();
    if ([VungleSDK sharedSDK].isAdPlayable) {
        [self.interstitialAdDelegate interstitialAdLoaded:self result:YES];
    } else {
        [self adLoad:1];
    }
}

- (void)interstitialAdShow {
    _DLOG();
    NSError *error = nil;
    if (![[VungleSDK sharedSDK] playAd:_viewController error:&error]) {
        _DLOG("%@", error);
        [self.interstitialAdDelegate interstitialAdClosed:self result:NO isSkipped:NO];
    }
}


#pragma - VungleSDKDelegate

- (void)vungleSDKwillShowAd {
    _DLOG();
    [self.interstitialAdDelegate interstitialAdShown:self];
}

- (void)vungleSDKWillCloseAdWithViewInfo:(NSDictionary *)viewInfo {
    _DLOG();
    NSNumber *completedView = [viewInfo objectForKey:@"completedView"];
    NSNumber *didDownload = [viewInfo objectForKey:@"didDownload"];
    
    if (didDownload.boolValue) {
        [self.interstitialAdDelegate interstitialAdClicked:self];
    }
    [self.interstitialAdDelegate interstitialAdClosed:self result:YES isSkipped:!completedView.boolValue];
}

- (void)vungleSDKAdPlayableChanged:(BOOL)isAdPlayable {
    _DLOG();
}


#pragma - Private methods

- (void)adLoad:(int)count {
    _DLOG("count=%d", count);
    
    dispatch_time_t time = dispatch_time(DISPATCH_TIME_NOW, 1 * NSEC_PER_SEC);
    dispatch_after(time, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        if ([VungleSDK sharedSDK].isAdPlayable) {
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.interstitialAdDelegate interstitialAdLoaded:self result:YES];
            });
        } else if (count == 5) {
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.interstitialAdDelegate interstitialAdLoaded:self result:NO];
            });
        } else {
            [self adLoad:count + 1];
        }
    });
}

@end

File: AdSwitcher-Android/vungle-adapter/src/main/java/net/adswitcher/adapter/vungle/VungleAdapter.java
package net.adswitcher.adapter.vungle;

import android.app.Activity;
import android.os.Handler;
import android.util.Log;

import com.vungle.publisher.EventListener;
import com.vungle.publisher.VunglePub;

import net.adswitcher.adapter.InterstitialAdAdapter;
import net.adswitcher.adapter.InterstitialAdListener;

import java.util.Map;

/**
 * Created by tkyaji on 2017/07/19.
 */

public class VungleAdapter implements InterstitialAdAdapter, EventListener {

    private static final String TAG = "VungleAdapter";

    private Activity activity;
    private InterstitialAdListener interstitialAdListener;


    @Override
    public void interstitialAdInitialize(Activity activity, InterstitialAdListener interstitialAdListener, Map<String, String> parameters, boolean testMode) {
        this.activity = activity;
        this.interstitialAdListener = interstitialAdListener;

        String appId = parameters.get("app_id");
        VunglePub.getInstance().init(activity, appId);
        VunglePub.getInstance().setEventListeners();
    }

    @Override
    public void interstitialAdLoad() {
        Log.d(TAG, "interstitialAdLoad");
        if (VunglePub.getInstance().isAdPlayable()) {
            this.interstitialAdListener.interstitialAdLoaded(this, true);
        } else {
            this.adLoad(1);
        }
    }

    @Override
    public void interstitialAdShow() {
        Log.d(TAG, "interstitialAdShow");
        VunglePub.getInstance().playAd();
    }


    @Override
    public void onAdEnd(boolean wasSuccessfulView, boolean wasCallToActionClicked) {
        Log.d(TAG, "onAdEnd");

        if (wasCallToActionClicked) {
            this.interstitialAdListener.interstitialAdClicked(this);
        }
        this.interstitialAdListener.interstitialAdClosed(this, true, !wasSuccessfulView);
    }

    @Override
    public void onAdStart() {
        Log.d(TAG, "onAdStart");
        this.interstitialAdListener.interstitialAdShown(this);
    }

    @Override
    public void onAdUnavailable(String reason) {
        Log.d(TAG, "onAdUnavailable : " + reason);
    }

    @Override
    public void onAdPlayableChanged(boolean isAdPlayable) {
        Log.d(TAG, "onAdPlayableChanged");
    }


    private void adLoad(final int count) {
        Log.d(TAG, "adLoad : count=" + count);

        (new Handler()).postDelayed(new Runnable() {
            @Override
            public void run() {
                if (VunglePub.getInstance().isAdPlayable()) {
                    VungleAdapter.this.activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            VungleAdapter.this.interstitialAdListener.interstitialAdLoaded(VungleAdapter.this, true);
                        }
                    });
                } else if (count == 5) {
                    VungleAdapter.this.activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            VungleAdapter.this.interstitialAdListener.interstitialAdLoaded(VungleAdapter.this, false);
                        }
                    });
                } else {
                    VungleAdapter.this.adLoad(count + 1);
                }
            }
        }, 1000);
    }
}
